#' The output of the differential expression analysis, used as the input
#' for function "extractSignificantGene" and "labelGenes"
#'
#' Generated by the differential expression analysis function based on DESeq2.
#' The data is a matrix that describes how different the expression levels of
#' genes are in different samples.
#'
#' @source Raw data obtained from Gene Expression Omnibus, generated by
#' GEAnaly::diffExpressionAnalysis.
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157103}{link}
#'
#' @format A matrix with 18309 rows and 6 columns, row names are gene names:
#' \describe{
#'  \item{baseMean}{The mean of normalized counts of all samples, normalizing
#'                  for sequencing depth}
#'  \item{log2FoldChange}{The effect size estimate, which tells about how much
#'                        the gene's expression seems to have changed in
#'                        different samples}
#'  \item{lfcSE}{Standard error of the log2FoldChange.}
#'  \item{stat}{log2FoldChange divided by lfcSE}
#'  \item{pvalue}{Use stat to compare to a standard Normal distribution to
#'                generate a two-tailed pvalue, which tells about whether the
#'                difference is significant}
#'  \item{padj}{Adjusted p value}
#' }
#'
#' @examples
#' \dontrun{
#'  diffExpressionResult
#' }
"diffExpressionResult"

#' The output of the enrichment analysis, used as the input
#' for function "visEnrichAnaly" and "visEnrichAnalyLollipop"
#'
#' A named list containing the named list of results from
#' gprofiler2 and a processed data frame based on the results from
#' gprofiler2 for visualization. Generated by the enrichment analysis function
#' based on gprofiler2.
#'
#' @source Raw data obtained from Gene Expression Omnibus, generated by
#' GEAnaly::enrichAnalysis.
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157103}{link}
#'
#' @format A named list of length 2, containing results of enrichment analysis:
#' \describe{
#'  \item{gProfilerResult}{A named list where 'result' contains a data frame
#'    with the enrichment analysis results and 'meta' contains metadata needed
#'    for visualization.}
#'  \item{enrichmentVis}{A processed data frame, which is a subset of the
#'    original enrichment analysis results, and also the geneRatio of different
#'    pathways are calculated and added for visualizations.}
#' }
#'
#' @examples
#' \dontrun{
#'  enrichOutputList
#' }
"enrichOutputList"

#' The gene expression matrix data for genes in different samples
#'
#' A data frame containing the matrix of un-normalized read
#' counts of gene expression data, normally obtained from RNA-seq or other
#' sequencing experiments. The count numbers in the matrix are
#' non-negative integers. The row names of the matrix are the genes,
#' and the column names are the sample IDs. This is a subset of
#' geneCountsDiffExpression, selecting only the first 105 genes and 41 samples,
#' since the original data is too large for the correlation analysis, which takes
#' too long for an example.
#'
#' @source Gene Expression Omnibus
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157103}{link}
#'
#' @format A matrix with 105 rows and 41 columns, row names are gene names,
#' column names are sample IDs:
#' \describe{
#'  \item{Column names}{Columns starting with C are sample with Covid. Samples
#'    starting with NC are samples without Covid}
#' }
#'
#' @examples
#' \dontrun{
#'  geneCountsCorrelation
#' }
"geneCountsCorrelation"

#' The gene expression matrix data for genes in different samples
#'
#' A data frame containing the matrix of un-normalized read
#' counts of gene expression data, normally obtained from RNA-seq or other
#' sequencing experiments. The count numbers in the matrix are
#' non-negative integers. The row names of the matrix are the genes,
#' and the column names are the sample IDs. Since the limit of the data size for
#' an R package, geneCountsDiffExpression is a subset of the original data.
#' geneCountsDiffExpression just uses the first 11000 genes.
#'
#' @source Gene Expression Omnibus
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157103}{link}
#'
#' @format A matrix with 11000 rows and 126 columns, row names are gene names,
#' column names are sample IDs:
#' \describe{
#'  \item{Column names}{Columns, whose name start with C, are samples with Covid.
#'    Columns, whose name start with NC, are samples without Covid.}
#' }
#'
#' @examples
#' \dontrun{
#'  geneCountsDiffExpression
#' }
"geneCountsDiffExpression"

#' A data frame that has the output of differential expression analysis and
#' label each gene with "UP", "DOWN" and "NOCHANGE"
#'
#' Basically similar to the diffExpressionResult, with an additional column
#' to label each gene. Genes that shows significant higher expression levels
#' are labelled with "UP", genes that shows significant lower expression
#' levels are labelled with "DOWN". Otherwise, genes are labelled
#' with "NOCHANGE".
#'
#' @source Raw data obtained from Gene Expression Omnibus, generated by
#' GEAnaly::diffExpressionAnalysis and GEAnaly::labelGenes.
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157103}{link}
#'
#' @format A matrix with 18309 rows and 7 columns, row names are gene names:
#' \describe{
#'  \item{baseMean}{The mean of normalized counts of all samples, normalizing
#'                  for sequencing depth}
#'  \item{log2FoldChange}{The effect size estimate, which tells about how much
#'                        the gene's expression seems to have changed in
#'                        different samples}
#'  \item{lfcSE}{Standard error of the log2FoldChange.}
#'  \item{stat}{log2FoldChange divided by lfcSE}
#'  \item{pvalue}{Use stat to compare to a standard Normal distribution to
#'                generate a two-tailed pvalue, which tells about whether the
#'                difference is significant}
#'  \item{padj}{Adjusted p value}
#'  \item{group}{Up-expressed genes are labelled with "UP", down-expressed genes
#'               are labelled with "DOWN". Otherwise, genes are labelled
#'               with "NOCHANGE".}
#' }
#'
#' @examples
#' \dontrun{
#'  labelledGenes
#' }
"labelledGenes"

#' A comparison table for sample IDs and sample conditions, supplementary
#' information for the gene count matrix used in differential expression analysis.
#'
#' A data frame containing the comparison table between
#' sample IDs and sample conditions, which are Covid and Noncovid.
#' The first column (sample) is the IDs or names of the samples, the second
#' column (condition) is the group that the sample belongs to.
#'
#' @source Gene Expression Omnibus.
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157103}{link}
#'
#' @format A data frame with 126 rows and 2 columns, row names are sample names:
#' \describe{
#'  \item{sample}{List all sample names}
#'  \item{condition}{Indicate the condition for each sample, with Covid or not}
#' }
#'
#' @examples
#' \dontrun{
#'  sampleInforDiffExpression
#' }
"sampleInforDiffExpression"

#' A data frame that is the subset of the output of differential expression
#' analysis, containing genes that have significantly different expression
#' levels in different samples.
#'
#' The data has the same columns as diffExpressionResult, but with fewer genes.
#' Genes in the data all have p values that are smaller than 0.05, and
#' absolute values of log2FoldChange that are larger than 2, which means that
#' these genes express significantly differently in different samples.
#'
#' @source Raw data obtained from Gene Expression Omnibus, generated by
#' GEAnaly::diffExpressionAnalysis and GEAnaly::extractSignificantGene.
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157103}{link}
#'
#' @format A data frame with 1299 rows and 6 columns, row names are sample names:
#' \describe{
#'  \item{baseMean}{The mean of normalized counts of all samples, normalizing
#'                  for sequencing depth}
#'  \item{log2FoldChange}{The effect size estimate, which tells about how much
#'                        the gene's expression seems to have changed in
#'                        different samples}
#'  \item{lfcSE}{Standard error of the log2FoldChange.}
#'  \item{stat}{log2FoldChange divided by lfcSE}
#'  \item{pvalue}{Use stat to compare to a standard Normal distribution to
#'                generate a two-tailed pvalue, which tells about whether the
#'                difference is significant}
#'  \item{padj}{Adjusted p value}
#' }
#'
#' @examples
#' \dontrun{
#'  significantGenes
#' }
"significantGenes"

#' A correlation matrix recording the correlation coefficient between genes
#'
#' A data frame contains the correlation coefficient of each pair of genes,
#' values on the diagonal are all zero.
#'
#' @source Raw data obtained from Gene Expression Omnibus, generated by
#' GEAnaly::corrAnalysis.
#' \href{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE157103}{link}
#'
#' @format A matrix with 103 rows and 103 columns, row names and column names
#' are same gene names. Each cell records the correlation coefficient of
#' corresponding genes on the x and y axis
#'
#' @examples
#' \dontrun{
#'  geneCorResult
#' }
"geneCorResult"

# [END]
